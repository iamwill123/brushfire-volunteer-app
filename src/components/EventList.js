import React, { useState } from 'react';
import Event from './Event';
import { SortEventListSelector } from './SortEventListSelector';

const EventList = ({ eventList }) => {
  const [inputVal, setinputVal] = useState('');

  const handleChange = event => {
    // If the search bar isn't empty
    if (event.target.value) {
      let value = event.target.value;
      setinputVal(value);
    } else {
      setinputVal('');
    }
  };

  if (!eventList) return 'Loading...';

  const filteredEventList = eventList.filter(event => {
    if (inputVal !== '') {
      const lcTitle = event.title.toLowerCase();
      const lcOrg = event.organization.toLowerCase();
      const filterVal = inputVal.toLowerCase();

      return lcTitle.includes(filterVal) || lcOrg.includes(filterVal);
    } else {
      return eventList;
    }
  });

  const FilterSearchInput = (
    <div className="page-divider">
      <div className="filter-search">
        <input
          type="text"
          className="input"
          value={inputVal}
          onChange={handleChange}
          placeholder="Search by org or title..."
        />
      </div>
    </div>
  );

  return (
    <>
      {FilterSearchInput}
      <SortEventListSelector />
      <div className="event-list">
        <h2 className="list-title">Volunteer in New York</h2>
        <div className="row" data-testid="event-list">
          {filteredEventList.map(event => (
            <Event key={event.id} event={event} />
          ))}
        </div>
      </div>
    </>
  );
};

export default EventList;
