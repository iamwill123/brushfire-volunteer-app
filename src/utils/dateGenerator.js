const options = {
  weekday: 'short',
  day: '2-digit',
  year: '2-digit',
  timeZone: 'America/New_York',
  hour12: true,
  hour: 'numeric',
  minute: 'numeric'
};

const dateTimeNow = () => new Date().toLocaleDateString('en-US', options);

const formatDateTime = dateString =>
  new Date(dateString).toLocaleDateString('en-US', options);

export { dateTimeNow, formatDateTime };
