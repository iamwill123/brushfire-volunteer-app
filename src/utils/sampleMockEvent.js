// 6 sample events
export const mockEventList = [
  {
    id: 0,
    title: 'Sports Instructor',
    host: 'Sam',
    createdAt: "2020-01-12T20:06:00.000Z",
    volunteerAt: null,
    startEndDate: 'Mon Feb 06, 2020 - Tue Jun 30, 2020',
    startEndTime: '1:00 PM - 9:00 PM',
    volunteer: false,
    organization: "I'RAISE Girls & Boys International",
    location: '495 3rd Ave. Brooklyn, NY 11215',
    imgURL: 'https://source.unsplash.com/ViJe3T1m5KE/800x450',
    desc: `<div class="event-desc"><p>We are looking for several Sports Instructors in the following sports: Martial Arts, Basketball, Football, Dance, with experience and passion for working with youth to volunteer his/her time for 3 hours a week at our Summer Camp Academy!</p></div>`,
    skills: [
      'Dance / Aerobics',
      'Baseball',
      'Martial Arts',
      'Dance',
      'Basketball'
    ],
    goodFor: ['Kids', 'Teens'],
    requirements: ['Background Check', '18 years +']
  },
  {
    id: 1,
    title: 'Creative Director',
    host: 'Buzz',
    createdAt: "2020-01-21T16:06:00.000Z",
    volunteerAt: null,
    startEndDate: 'Mon Feb 16, 2020 - Tue July 10, 2020',
    startEndTime: '12:00 PM - 2:00 PM',
    volunteer: false,
    organization: 'Hosh Yoga',
    location: '55 Nassau Ave. Brooklyn, NY 11222',
    imgURL: 'https://source.unsplash.com/4H67yWXO2-M/800x450',
    desc: `<div class="event-desc">
      <p><strong>Our Mission</strong></p><p><em>Enrichment education as a right of life rather than a luxury.</em></p><p>Hosh is here to provide a fair chance to explore healthier living for children through yoga, dance, meditation, workshops and strong community development in the Greenpoint and Williamsburg areas of Brooklyn, New York and beyond.</p><p>We invite volunteers to support us in any of the following area:</p><p>Creative Director</p><p>Photographer/Videographer</p><p>Hours: 3-6 hours per week. 2-3 days a week. Some remote work possible.</p><p>Join us! Email us today Info@hoshkids.org</p>
    </div>`,
    skills: [
      'Tutoring',
      'Mentoring',
      'Video editing',
      'Photography',
      'Creativity'
    ],
    goodFor: ['Kids', 'Teens'],
    requirements: ['Background Check', '18 years +']
  },
  {
    id: 2,
    title:
      "Change Kids' Lives",
    host: 'Millie',
    createdAt: "2019-09-02T03:06:00.000Z",
    volunteerAt: null,
    startEndDate: 'Mon Mar 01, 2020 - Fri Jun 30, 2020',
    startEndTime: '4:00 PM - 6:00 PM',
    volunteer: false,
    organization: 'Hosh Yoga',
    location: '55 Nassau Ave. Brooklyn, NY 11222',
    imgURL: 'https://source.unsplash.com/bTp32-t6K7c/800x450',
    desc: `<div class="event-desc"><p><strong>Use your passion, skills and experience to teach enrichment education in an NYC public school.<br></strong></p><br>Make a difference in a low income community, spreading our mission and growing your teaching skills--teaching stipends will be provided for all. Arts on a budget! Enrichment education for children as a right of life rather than a luxury is our mission. Join our vision to bring enrichment education to all children as we work with over 40 schools and over 6,000 school children. <p>Our <strong>teaching fellowship</strong> involves an stipended 12 week opportunity to teach in a local public pre-school, elementary, or middle school enrichment program once a week.</p></div>`,
    skills: [
      'Physical Education',
      'Community Outreach',
      'Teaching / Instruction',
      'Sports Coaching'
    ],
    goodFor: ['Adults', 'Teens'],
    requirements: ['Must be at least 20', 'Orientation or Training', '6 months']
  },
  {
    id: 3,
    title: 'Physical Activity Coach',
    host: 'Stephanie',
    createdAt: "2020-01-01T19:22:00.000Z",
    volunteerAt: null,
    startEndDate: 'Mon Jan 06, 2020 - Fri Sept 30, 2020',
    startEndTime: '1:00 PM - 2:00 PM',
    volunteer: false,
    organization: 'POSITIVE TRANSITIONING INC',
    location: 'Virtual Opportunity',
    imgURL: 'https://source.unsplash.com/mse1vdzZXjA/800x450',
    desc: `<div class="event-desc"><ul><li>Create daily workout regimes</li><li>Host online walking meetings</li><li>Create mid day quick workout challenges</li><li>Create other physical activity ideas</li></ul></div>`,
    skills: [
      'Physical Education',
      'Community Outreach',
      'Teaching / Instruction',
      'Sports Coaching'
    ],
    goodFor: ['Adults', 'Teens'],
    requirements: ['Must be at least 20', 'Orientation or Training']
  },
  {
    id: 4,
    title: 'Prayer Line Volunteer',
    host: 'Billy',
    createdAt: "2019-01-01T14:06:00.000Z",
    volunteerAt: null,
    startEndDate: 'Mon Jan 01, 2020 - Fri Sept 30, 2020',
    startEndTime: '3:00 PM - 5:00 PM',
    volunteer: false,
    organization: 'POSITIVE TRANSITIONING INC',
    location: 'Virtual Opportunity',
    imgURL: 'https://source.unsplash.com/rzmQOng8h8I/800x450',
    desc: `<div class="event-desc"><p>Philippians 2:3-4 Do nothing from selfishness or empty conceit, but with humility of mind regard one another as more important than yourselves; do not merely look out for your own personal interests, but also for the interests of others. Greetings! I am Karen Anne. I am just like you that have a heart to serve for the Lord. In serving the Lord it is one of the Lord's expectation for us to pray for one another. I am calling on all Prayer Intercessors to join me in our prayer room for Prayer Training on 01/20/2019 @2:00pm EST. The Prayer Intercessors that I am seeking would have the following gifts.</p> <ol><li> The willingness to listen with empathy.</li> <li> The gift of prayer</li> <li> A clear speaking voice </li> <li> The ability to be humble </li> <li> The Lord's power while praying</li> <li> Tremendous faith As, you are being led by the Holy Spirit </li> </ol><p>I ask that your prayerfully consider attending the training. Also, before the training I would love to get to know you. I want you to get to know me. When you sign up for the class before the class start I welcome you contacting me. 313.212.9284. Thank you for your time! I look forward to chatting with you. Yours faithfully,
    Karen Anne~~</p></div>`,
    skills: [
      'Networking',
      'People Skills',
      'Public Speaking',
      'Relationship Building',
      'Verbal / Written Communication'
    ],
    goodFor: ['All'],
    requirements: ['Orientation or Training']
  },
  {
    id: 5,
    title: 'Tutoring & Mentoring',
    host: 'Satoshi',
    createdAt: "2020-01-16T08:06:00.000Z",
    volunteerAt: null,
    startEndDate: 'Mon Mar 11, 2020 - Tue Oct 30, 2020',
    startEndTime: '3:00 PM - 5:00 PM',
    volunteer: false,
    organization: 'Youth Potential Academy',
    location: '10 Grand Army Plaza. Brooklyn, NY 11238',
    imgURL: 'https://source.unsplash.com/5fNmWej4tAA/800x450',
    desc: `<div class="event-desc"><p>Are you looking to be apart of an organization that passionately cares in making a difference within the youth community? Youth Potential Academy recognizes that each youth is different and may learn differently. This is why our curriculum is individualized, student centered, and creatively delivered to allow for optimal learning. We align academics with the students individual passions. Each student we serve at Youth Potential Academy is assessed and matched with professional and qualified instructors who are passionate about guiding the student towards greater academic and personal growth. <br><br>Student age group (10-17)<br>Travel reimbursement available.  </p><ul><li>On-site Location</li><li>Stipend Provided</li><li>Training Provided</li><li>Groups Welcome</li><li>International Volunteers Welcome</li></ul>  If interested apply today to aid the next generation.<p></p>`,
    skills: [
      'Tutoring',
      'Mentoring',
      'Math / Science Instruction',
      'Reading / Writing',
      'STEM',
      'Mathematics'
    ],
    goodFor: ['Kids', 'Teens'],
    requirements: ['Background Check', '18 years +']
  }
];
