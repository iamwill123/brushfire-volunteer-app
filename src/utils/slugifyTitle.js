import slugify from 'slugify';

export const slugifyTitle = (value = '') => {
  if (typeof value !== 'string') {
    throw new Error('Requires value to be type String.');
  }
  return slugify(value, {
    replacement: '-', // replace spaces with '-'
    remove: /[,/*+~.()'"!:@]/g, // regex to remove characters
    lower: true // result in lower case
  });
};

