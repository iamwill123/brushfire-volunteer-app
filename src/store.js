import React, { createContext, useReducer } from 'react';
import { dateTimeNow } from './utils/dateGenerator';
import { mockEventList } from './utils/sampleMockEvent';

const initialState = mockEventList; // sample events
const store = createContext(initialState);
const { Provider } = store;

function init(initialState) {
  return initialState;
}

const reducer = (state, action) => {
  switch (action.type) {
    case 'GET_EVENTS': {
      return state;
    }

    case 'TOGGLE_VOLUNTEER': {
      const newState = state.map(event =>
        event.id === action.id
          ? {
              ...event,
              volunteer: !event.volunteer,
              volunteerAt: !event.volunteer ? dateTimeNow() : null
            }
          : event
      );
      return newState;
    }

    case 'SORT_BY': {
      const { sortBy = '' } = action;

      if (sortBy === 'most recent') {
        // sort by most recent date
        const byMostRecent = (a, b) => {
          if (a.createdAt < b.createdAt) return 1;
          if (a.createdAt > b.createdAt) return -1;
          return 0;
        };
        const newState = state.slice().sort(byMostRecent);
        return newState;
      } else if (sortBy === 'organization') {
        // sort by alphabetical order
        const byOrganization = (a, b) => {
          if (a.organization < b.organization) return -1;
          if (a.organization > b.organization) return 1;
          return 0;
        };
        const newState = state.slice().sort(byOrganization);
        return newState;
      } else {
        return state;
      }
    }

    case 'RESET':
      return init(initialState);

    default:
      throw new Error(`Unhandled action type: ${action.type}`);
  }
};

const StateProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState, init);
  return <Provider value={{ state, dispatch }}>{children}</Provider>;
};

export { store, StateProvider };
