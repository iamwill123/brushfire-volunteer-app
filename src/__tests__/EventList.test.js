import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, waitForElement, fireEvent } from '@testing-library/react';
import EventList from '../components/EventList';
import { mockEventList } from '../utils/sampleMockEvent';
import { StateProvider } from '../store';

// https://testing-library.com/docs/react-testing-library/intro
describe('<EventList /> component', () => {
  let list;

  beforeEach(() => {
    list = render(
      <StateProvider>
        <Router>
          <EventList eventList={mockEventList} />
        </Router>
      </StateProvider>
    );
  });

  it('matches snapshot', () => {
    const { asFragment } = list;
    expect(asFragment()).toMatchSnapshot();
  });

  // As a user I should see a list of events
  test('should render correct number of events', async () => {
    const { getByTestId } = list;

    const eventList = await waitForElement(() => getByTestId('event-list'));
    expect(eventList.childElementCount).toBe(6);
    expect(eventList.classList[0]).toBe('row');
  });

  test('should toggle volunteer input', async () => {
    const { getByTestId, getByText } = list;

    // as a user I can select an event I want to volunteer for
    const volunteerToggleElement = getByTestId('volunteer-toggle-0');
    expect(volunteerToggleElement.innerHTML).toBe('Volunteer');

    fireEvent.click(volunteerToggleElement);
    expect(getByText('Volunteering')).toBeInTheDocument();
    expect(volunteerToggleElement.innerHTML).toBe('Volunteering');

    // as a user I can de-select an event I don't want to volunteer for
    const volunteerToggleElement2 = getByTestId('volunteer-toggle-2');
    expect(volunteerToggleElement2.innerHTML).toBe('Volunteer');
    // toggle
    fireEvent.click(volunteerToggleElement2);
    expect(volunteerToggleElement2.innerHTML).toBe('Volunteering');
    fireEvent.click(volunteerToggleElement2);
    expect(volunteerToggleElement2.innerHTML).toBe('Volunteer');
  });
});

