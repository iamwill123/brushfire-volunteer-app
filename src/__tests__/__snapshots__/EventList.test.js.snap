// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventList /> component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="page-divider"
  >
    <div
      class="filter-search"
    >
      <input
        class="input"
        placeholder="Search by org or title..."
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    class="sort-selector"
  >
    <span>
      <input
        readonly=""
        type="checkbox"
      />
      <span>
        most recent
      </span>
    </span>
    <span>
      <input
        readonly=""
        type="checkbox"
      />
      <span>
        organization
      </span>
    </span>
  </div>
  <div
    class="event-list"
  >
    <h2
      class="list-title"
    >
      Volunteer in New York
    </h2>
    <div
      class="row"
      data-testid="event-list"
    >
      <div
        class="event-card"
      >
        <a
          href="/event/sports-instructor"
        >
          <img
            alt="Sports Instructor"
            src="https://source.unsplash.com/ViJe3T1m5KE/800x450"
          />
        </a>
        <div
          class="event-volunteer-btn"
        >
          <span>
            <input
              readonly=""
              type="checkbox"
            />
            <span
              data-testid="volunteer-toggle-0"
            >
              Volunteer
            </span>
          </span>
        </div>
        <div
          class="event-desc"
        >
          <div
            class="event-date"
            data-testid="event-date"
          >
            Mon Feb 06, 2020 - Tue Jun 30, 2020
          </div>
          <div
            data-testid="event-time"
          >
            1:00 PM - 9:00 PM
          </div>
          <a
            href="/event/sports-instructor"
          >
            <h2
              data-testid="event-title"
            >
              Sports Instructor
            </h2>
          </a>
          <sub>
            with I'RAISE Girls & Boys International
          </sub>
          <p
            class="event-location"
            data-testid="event-location"
          >
            495 3rd Ave. Brooklyn, NY 11215
          </p>
        </div>
      </div>
      <div
        class="event-card"
      >
        <a
          href="/event/creative-director"
        >
          <img
            alt="Creative Director"
            src="https://source.unsplash.com/4H67yWXO2-M/800x450"
          />
        </a>
        <div
          class="event-volunteer-btn"
        >
          <span>
            <input
              readonly=""
              type="checkbox"
            />
            <span
              data-testid="volunteer-toggle-1"
            >
              Volunteer
            </span>
          </span>
        </div>
        <div
          class="event-desc"
        >
          <div
            class="event-date"
            data-testid="event-date"
          >
            Mon Feb 16, 2020 - Tue July 10, 2020
          </div>
          <div
            data-testid="event-time"
          >
            12:00 PM - 2:00 PM
          </div>
          <a
            href="/event/creative-director"
          >
            <h2
              data-testid="event-title"
            >
              Creative Director
            </h2>
          </a>
          <sub>
            with Hosh Yoga
          </sub>
          <p
            class="event-location"
            data-testid="event-location"
          >
            55 Nassau Ave. Brooklyn, NY 11222
          </p>
        </div>
      </div>
      <div
        class="event-card"
      >
        <a
          href="/event/change-kids-lives"
        >
          <img
            alt="Change Kids' Lives"
            src="https://source.unsplash.com/bTp32-t6K7c/800x450"
          />
        </a>
        <div
          class="event-volunteer-btn"
        >
          <span>
            <input
              readonly=""
              type="checkbox"
            />
            <span
              data-testid="volunteer-toggle-2"
            >
              Volunteer
            </span>
          </span>
        </div>
        <div
          class="event-desc"
        >
          <div
            class="event-date"
            data-testid="event-date"
          >
            Mon Mar 01, 2020 - Fri Jun 30, 2020
          </div>
          <div
            data-testid="event-time"
          >
            4:00 PM - 6:00 PM
          </div>
          <a
            href="/event/change-kids-lives"
          >
            <h2
              data-testid="event-title"
            >
              Change Kids' Lives
            </h2>
          </a>
          <sub>
            with Hosh Yoga
          </sub>
          <p
            class="event-location"
            data-testid="event-location"
          >
            55 Nassau Ave. Brooklyn, NY 11222
          </p>
        </div>
      </div>
      <div
        class="event-card"
      >
        <a
          href="/event/physical-activity-coach"
        >
          <img
            alt="Physical Activity Coach"
            src="https://source.unsplash.com/mse1vdzZXjA/800x450"
          />
        </a>
        <div
          class="event-volunteer-btn"
        >
          <span>
            <input
              readonly=""
              type="checkbox"
            />
            <span
              data-testid="volunteer-toggle-3"
            >
              Volunteer
            </span>
          </span>
        </div>
        <div
          class="event-desc"
        >
          <div
            class="event-date"
            data-testid="event-date"
          >
            Mon Jan 06, 2020 - Fri Sept 30, 2020
          </div>
          <div
            data-testid="event-time"
          >
            1:00 PM - 2:00 PM
          </div>
          <a
            href="/event/physical-activity-coach"
          >
            <h2
              data-testid="event-title"
            >
              Physical Activity Coach
            </h2>
          </a>
          <sub>
            with POSITIVE TRANSITIONING INC
          </sub>
          <p
            class="event-location"
            data-testid="event-location"
          >
            Virtual Opportunity
          </p>
        </div>
      </div>
      <div
        class="event-card"
      >
        <a
          href="/event/prayer-line-volunteer"
        >
          <img
            alt="Prayer Line Volunteer"
            src="https://source.unsplash.com/rzmQOng8h8I/800x450"
          />
        </a>
        <div
          class="event-volunteer-btn"
        >
          <span>
            <input
              readonly=""
              type="checkbox"
            />
            <span
              data-testid="volunteer-toggle-4"
            >
              Volunteer
            </span>
          </span>
        </div>
        <div
          class="event-desc"
        >
          <div
            class="event-date"
            data-testid="event-date"
          >
            Mon Jan 01, 2020 - Fri Sept 30, 2020
          </div>
          <div
            data-testid="event-time"
          >
            3:00 PM - 5:00 PM
          </div>
          <a
            href="/event/prayer-line-volunteer"
          >
            <h2
              data-testid="event-title"
            >
              Prayer Line Volunteer
            </h2>
          </a>
          <sub>
            with POSITIVE TRANSITIONING INC
          </sub>
          <p
            class="event-location"
            data-testid="event-location"
          >
            Virtual Opportunity
          </p>
        </div>
      </div>
      <div
        class="event-card"
      >
        <a
          href="/event/tutoring-and-mentoring"
        >
          <img
            alt="Tutoring & Mentoring"
            src="https://source.unsplash.com/5fNmWej4tAA/800x450"
          />
        </a>
        <div
          class="event-volunteer-btn"
        >
          <span>
            <input
              readonly=""
              type="checkbox"
            />
            <span
              data-testid="volunteer-toggle-5"
            >
              Volunteer
            </span>
          </span>
        </div>
        <div
          class="event-desc"
        >
          <div
            class="event-date"
            data-testid="event-date"
          >
            Mon Mar 11, 2020 - Tue Oct 30, 2020
          </div>
          <div
            data-testid="event-time"
          >
            3:00 PM - 5:00 PM
          </div>
          <a
            href="/event/tutoring-and-mentoring"
          >
            <h2
              data-testid="event-title"
            >
              Tutoring & Mentoring
            </h2>
          </a>
          <sub>
            with Youth Potential Academy
          </sub>
          <p
            class="event-location"
            data-testid="event-location"
          >
            10 Grand Army Plaza. Brooklyn, NY 11238
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
